src/zenml/logging/step_logging.py

```diff
@@ -635,70 +695,47 @@ def __exit__(
             exc_val: The instance of the exception
             exc_tb: The traceback of the exception
 
-        Restores the `write` method of both stderr and stdout.
+        Removes the handler from loggers and context variables.
         """
-        # Step 1: Restore stdout/stderr FIRST to prevent logging during shutdown
-        try:
-            if self._original_methods_saved:
-                setattr(sys.stdout, "write", self.stdout_write)
-                setattr(sys.stderr, "write", self.stderr_write)
-                redirected.set(False)
-        except Exception:
-            pass
+        if exc_type is not None:
+            # Write the exception and its traceback to the logs
+            self.artifact_store_handler.emit(
+                logging.LogRecord(
+                    name="exception",
+                    level=logging.ERROR,
+                    pathname="",
+                    lineno=0,
+                    msg="An exception has occurred.",
+                    args=(),
+                    exc_info=(exc_type, exc_val, exc_tb) if exc_val else None,
+                )
+            )
 
-        # Step 2: Shutdown thread (it will automatically drain queue and merge files)
+        # Remove handler from root logger and restore original level
+        if self.artifact_store_handler:
+            root_logger = logging.getLogger()
+            # Check if handler is still in the root logger before removing
+            if self.artifact_store_handler in root_logger.handlers:
+                root_logger.removeHandler(self.artifact_store_handler)
+            # Restore original root logger level
+            if hasattr(self, "original_root_level"):
+                root_logger.setLevel(self.original_root_level)
+
+        # Remove handler from context variables
+        handlers = logging_handlers.get().copy()
+        if self.artifact_store_handler in handlers:
+            handlers.remove(self.artifact_store_handler)
+        logging_handlers.set(handlers)
```

Comment:
This is a potential cause for race conditions. Multitple threads could be removing their own handler in parallel, which would cause some handler to be left over even if all should have been removed.

Reply:
I have implemented a `ContextVarList` that protects it against race conditions while `append`ing and `remove`ing with a `lock`.